PR: https://github.com/odoo/odoo/pull/231159

From: 9605045313953b4c8c734c0d52e8032e3c36bf3a
From: Krzysztof Magusiak (krma)
Date: 2025-11-28 09:30:18

Breaking data model changes scores: del:27 + add:0, change matches:
-    job_title = fields.Char(related='employee_id.job_title')
-    work_phone = fields.Char(related='employee_id.work_phone', readonly=False, related_sudo=False)
-    mobile_phone = fields.Char(related='employee_id.mobile_phone', readonly=False, related_sudo=False)
-    work_email = fields.Char(related='employee_id.work_email', readonly=False, related_sudo=False)
-    category_ids = fields.Many2many(related='employee_id.category_ids', string="Employee Tags", readonly=False, related_sudo=False)
-    work_contact_id = fields.Many2one(related='employee_id.work_contact_id', readonly=False, related_sudo=False)
-    work_location_id = fields.Many2one(related='employee_id.work_location_id', readonly=False, related_sudo=False)
-    private_street = fields.Char(related='employee_id.private_street', string="Private Street", readonly=False, related_sudo=False)
-    private_street2 = fields.Char(related='employee_id.private_street2', string="Private Street2", readonly=False, related_sudo=False)
-    private_city = fields.Char(related='employee_id.private_city', string="Private City", readonly=False, related_sudo=False)
-    private_state_id = fields.Many2one(
-    private_zip = fields.Char(related='employee_id.private_zip', readonly=False, string="Private Zip", related_sudo=False)
-    private_country_id = fields.Many2one(related='employee_id.private_country_id', string="Private Country", readonly=False, related_sudo=False)
-    private_phone = fields.Char(related='employee_id.private_phone', readonly=False, related_sudo=False)
-    private_email = fields.Char(related='employee_id.private_email', string="Private Email", readonly=False)
-    km_home_work = fields.Integer(related='employee_id.km_home_work', readonly=False, related_sudo=False)
-    employee_bank_account_ids = fields.Many2many('res.partner.bank', related='employee_id.bank_account_ids', string="Employee's Bank Accounts", related_sudo=False, readonly=False)
-    emergency_contact = fields.Char(related='employee_id.emergency_contact', readonly=False, related_sudo=False)
-    emergency_phone = fields.Char(related='employee_id.emergency_phone', readonly=False, related_sudo=False)
-    visa_expire = fields.Date(related='employee_id.visa_expire', readonly=False, related_sudo=False)
-    additional_note = fields.Text(related='employee_id.additional_note', readonly=False, related_sudo=False)
-    barcode = fields.Char(related='employee_id.barcode', readonly=False, related_sudo=False)
-    pin = fields.Char(related='employee_id.pin', readonly=False, related_sudo=False)
-    bank_account_ids = fields.Many2many(related="employee_id.bank_account_ids")

Total Changes: 207

[IMP] hr: user employee fields

Most of the fields added on the res.users model come directly from the
hr.employee model, which is accessible only if the user has HR
permissions. This is not the case for most users, and therefore we need
to read such fields in sudo, but only these fields.

We replace the existing related fields by an equivalent computed field.
The `field_employee` function is a builder for such fields. It adds the
right compute, inverse and search methods.

**Field updates:**
* employee_ids: dropped, seems only used for invalidation
* employee_id: computed, adding security to search using sudo only when
  we compute the value for the current user
* active: already accessible
* child_ids: already accessible
* employee_resource_calendar_id: sudo related field
* work_contact_id: employee field (not writeable)
* bank_account_ids: employee field (not writeable)
* is_hr_user: current user info
* is_system: current user info
* additional_note: employee field
* barcode: employee field
* category_ids: employee field
* display_name: dropped, no inverse
* emergency_contact: employee field
* employee_bank_account_ids: employee field (deprecated?)
* job_title: employee field
* km_home_work: employee field
* mobile_phone: employee field
* pin: employee field
* private_*: employee fields (address information)
* visa_expire: employee field
* work_email: employee field
* work_location_id: employee field, name and type and related_sudo
* work_phone: employee field

Removing view hacks since normal permissions are implemented.

Fix tests to assign `employee_ids` instead of `employee_id` as it is a
context-dependent computed field based on all ids.

Part-of: odoo/odoo#231159
Related: odoo/enterprise#96931
Related: odoo/upgrade#8625
Signed-off-by: Raphael Collet <rco@odoo.com>

================================= pseudo patch: =================================

--- a/addons/hr/models/res_users.py
+++ b/addons/hr/models/res_users.py
@@ -1,54 +1,69 @@
-# -*- coding: utf-8 -*-
 # Part of Odoo. See LICENSE file for full copyright and licensing details.
 
 import ast
 
 from markupsafe import Markup
 
-from odoo import api, models, fields, _, SUPERUSER_ID
+from odoo import _, api, fields, models
 from odoo.exceptions import AccessError
 from odoo.fields import Domain
 from odoo.tools.misc import clean_context
-from odoo.addons.mail.tools.discuss import Store
-
-
-HR_READABLE_FIELDS = [
-    'active',
-    'child_ids',
-    'employee_id',
-    'employee_ids',
-    'is_hr_user',
-    'is_system',
-    'employee_resource_calendar_id',
-    'work_contact_id',
-    'bank_account_ids',
-]
-
-HR_WRITABLE_FIELDS = [
-    'additional_note',
-    'private_street',
-    'private_street2',
-    'private_city',
-    'private_state_id',
-    'private_zip',
-    'private_country_id',
-    'private_phone',
-    'private_email',
-    'barcode',
-    'category_ids',
-    'display_name',
-    'emergency_contact',
-    'emergency_phone',
-    'employee_bank_account_ids',
-    'job_title',
-    'km_home_work',
-    'mobile_phone',
-    'pin',
-    'visa_expire',
-    'work_email',
-    'work_location_id',
-    'work_phone',
-]
+
+
+def field_employee(field_type: type[fields.Field], name: str, *, field_name='', user_writeable=False, **kw):
+    """Simulating a related field "employee_id.{name}".
+
+    Current user is read and updated with `sudo()`. Otherwise, behaves as a
+    simple related field.
+
+    :param field_name: provide if you name your field in some other way than
+        on the employee model
+    """
+    # simulating a related field which bypasses access only for the current user
+    assert 'related' not in kw and 'store' not in kw, "Unsupported parameters found"
+    if not field_name:
+        field_name = name
+
+    @api.depends(f'employee_id.{name}')
+    @api.depends_context('uid')
+    def compute_employee_field(self):
+        current_user = self.env.user
+        for user in self:
+            if user == current_user:
+                employee = user.sudo().employee_id.with_prefetch()
+            else:
+                employee = user.employee_id
+            user[field_name] = employee[name]
+
+    def inverse_employee_field(self):
+        current_user = self.env.user
+        for user in self:
+            if user_writeable and user == current_user:
+                employee = user.sudo().employee_id.with_prefetch()
+            else:
+                employee = user.employee_id
+            employee[name] = user[field_name]
+
+    def search_employee_field(self, operator, value):
+        if operator in Domain.NEGATIVE_OPERATORS:
+            return NotImplemented
+        subdomain = Domain(name, operator, value)
+        Employee = self.env['hr.employee']
+        if Employee.has_access('read') and Employee._has_field_access(Employee._fields[name], 'read'):
+            return Domain('employee_id', 'any', subdomain)
+        else:
+            user = self.env.user.employee_id.filtered_domain(subdomain).user_id
+            return Domain('id', 'in', user.ids)
+
+    # mark the compute function for identification later
+    compute_employee_field.is_field_employee = True
+    return field_type(
+        **kw,
+        compute=compute_employee_field,
+        inverse=inverse_employee_field,
+        search=search_employee_field,
+        user_writeable=user_writeable,
+    )
 
 
 class ResUsers(models.Model):
@@ -57,46 +72,46 @@ class ResUsers(models.Model):
     def _employee_ids_domain(self):
         # employee_ids is considered a safe field and as such will be fetched as sudo.
         # So try to enforce the security rules on the field to make sure we do not load employees outside of active companies
-        return [('company_id', 'in', self.env.company.ids + self.env.context.get('allowed_company_ids', []))]
+        return [('company_id', 'in', self.env.companies.ids)]
 
     # note: a user can only be linked to one employee per company (see sql constraint in `hr.employee`)
     employee_ids = fields.One2many('hr.employee', 'user_id', string='Related employee', domain=_employee_ids_domain)
     employee_public_ids = fields.One2many('hr.employee.public', 'user_id', string='Related employee (public)', domain=_employee_ids_domain, readonly=True)
     employee_id = fields.Many2one('hr.employee', string="Company employee",
-        compute='_compute_company_employee', search='_search_company_employee', store=False)
-
-    job_title = fields.Char(related='employee_id.job_title')
-    work_phone = fields.Char(related='employee_id.work_phone', readonly=False, related_sudo=False)
-    mobile_phone = fields.Char(related='employee_id.mobile_phone', readonly=False, related_sudo=False)
-    work_email = fields.Char(related='employee_id.work_email', readonly=False, related_sudo=False)
-    category_ids = fields.Many2many(related='employee_id.category_ids', string="Employee Tags", readonly=False, related_sudo=False)
-    work_contact_id = fields.Many2one(related='employee_id.work_contact_id', readonly=False, related_sudo=False)
-    work_location_id = fields.Many2one(related='employee_id.work_location_id', readonly=False, related_sudo=False)
+        compute='_compute_company_employee', search='_search_company_employee', readonly=True)
+
+    job_title = field_employee(fields.Char, 'job_title', user_writeable=True)
+    work_phone = field_employee(fields.Char, 'work_phone', user_writeable=True)
+    mobile_phone = field_employee(fields.Char, 'mobile_phone', user_writeable=True)
+    work_email = field_employee(fields.Char, 'work_email', user_writeable=True)
+    category_ids = field_employee(fields.Many2many, 'category_ids', comodel_name='hr.employee.category', string="Employee Tags", user_writeable=True)
+    work_contact_id = field_employee(fields.Many2one, 'work_contact_id', comodel_name='res.partner')
+    work_location_id = field_employee(fields.Many2one, 'work_location_id', comodel_name='hr.work.location', user_writeable=True)
     work_location_name = fields.Char(related="employee_id.work_location_name")
     work_location_type = fields.Selection(related="employee_id.work_location_type")
-    private_street = fields.Char(related='employee_id.private_street', string="Private Street", readonly=False, related_sudo=False)
-    private_street2 = fields.Char(related='employee_id.private_street2', string="Private Street2", readonly=False, related_sudo=False)
-    private_city = fields.Char(related='employee_id.private_city', string="Private City", readonly=False, related_sudo=False)
-    private_state_id = fields.Many2one(
-        related='employee_id.private_state_id', string="Private State", readonly=False, related_sudo=False,
+    private_street = field_employee(fields.Char, 'private_street', string="Private Street", user_writeable=True)
+    private_street2 = field_employee(fields.Char, 'private_street2', string="Private Street2", user_writeable=True)
+    private_city = field_employee(fields.Char, 'private_city', string="Private City", user_writeable=True)
+    private_state_id = field_employee(fields.Many2one, 'private_state_id', string="Private State", user_writeable=True,
+        comodel_name='res.country.state',
         domain="[('country_id', '=?', private_country_id)]")
-    private_zip = fields.Char(related='employee_id.private_zip', readonly=False, string="Private Zip", related_sudo=False)
-    private_country_id = fields.Many2one(related='employee_id.private_country_id', string="Private Country", readonly=False, related_sudo=False)
-    private_phone = fields.Char(related='employee_id.private_phone', readonly=False, related_sudo=False)
-    private_email = fields.Char(related='employee_id.private_email', string="Private Email", readonly=False)
-    km_home_work = fields.Integer(related='employee_id.km_home_work', readonly=False, related_sudo=False)
+    private_zip = field_employee(fields.Char, 'private_zip', string="Private Zip", user_writeable=True)
+    private_country_id = field_employee(fields.Many2one, 'private_country_id', comodel_name='res.country', string="Private Country", user_writeable=True)
+    private_phone = field_employee(fields.Char, 'private_phone', user_writeable=True)
+    private_email = field_employee(fields.Char, 'private_email', user_writeable=True)
+    km_home_work = field_employee(fields.Integer, 'km_home_work', user_writeable=True)
     # res.users already have a field bank_account_id and country_id from the res.partner inheritance: don't redefine them
     # This field no longer appears to be in use. To avoid breaking anything it must only be removed after the freeze of v19.
-    employee_bank_account_ids = fields.Many2many('res.partner.bank', related='employee_id.bank_account_ids', string="Employee's Bank Accounts", related_sudo=False, readonly=False)
-    emergency_contact = fields.Char(related='employee_id.emergency_contact', readonly=False, related_sudo=False)
-    emergency_phone = fields.Char(related='employee_id.emergency_phone', readonly=False, related_sudo=False)
-    visa_expire = fields.Date(related='employee_id.visa_expire', readonly=False, related_sudo=False)
-    additional_note = fields.Text(related='employee_id.additional_note', readonly=False, related_sudo=False)
-    barcode = fields.Char(related='employee_id.barcode', readonly=False, related_sudo=False)
-    pin = fields.Char(related='employee_id.pin', readonly=False, related_sudo=False)
+    employee_bank_account_ids = field_employee(fields.Many2many, 'bank_account_ids', comodel_name='res.partner.bank', string="Employee's Bank Accounts", user_writeable=True, field_name='employee_bank_account_ids')
+    emergency_contact = field_employee(fields.Char, 'emergency_contact', user_writeable=True)
+    emergency_phone = field_employee(fields.Char, 'emergency_phone', user_writeable=True)
+    visa_expire = field_employee(fields.Date, 'visa_expire', user_writeable=True)
+    additional_note = field_employee(fields.Text, 'additional_note', user_writeable=True)
+    barcode = field_employee(fields.Char, 'barcode', user_writeable=True)
+    pin = field_employee(fields.Char, 'pin', user_writeable=True)
     employee_count = fields.Integer(compute='_compute_employee_count')
     employee_resource_calendar_id = fields.Many2one(related='employee_id.resource_calendar_id', string="Employee's Working Hours", readonly=True)
-    bank_account_ids = fields.Many2many(related="employee_id.bank_account_ids")
+    bank_account_ids = field_employee(fields.Many2many, 'bank_account_ids', comodel_name='res.partner.bank')
 
     create_employee = fields.Boolean(store=False, default=False, copy=False, string="Technical field, whether to create an employee")
     create_employee_id = fields.Many2one('hr.employee', store=False, copy=False, string="Technical field, bind user to this employee on create")
@@ -108,59 +123,21 @@ class ResUsers(models.Model):
     def _compute_is_system(self):
         self.is_system = self.env.user._is_system()
 
+    @api.depends_context('uid')
     def _compute_is_hr_user(self):
         is_hr_user = self.env.user.has_group('hr.group_hr_user')
-        for user in self:
-            user.is_hr_user = is_hr_user
+        self.is_hr_user = is_hr_user
 
     @api.depends('employee_ids')
     def _compute_employee_count(self):
         for user in self.with_context(active_test=False):
             user.employee_count = len(user.employee_ids)
 
-    @property
-    def SELF_READABLE_FIELDS(self):
-        return super().SELF_READABLE_FIELDS + HR_READABLE_FIELDS + HR_WRITABLE_FIELDS
-
-    @property
-    def SELF_WRITEABLE_FIELDS(self):
-        return super().SELF_WRITEABLE_FIELDS + HR_WRITABLE_FIELDS
-
     @api.onchange("private_state_id")
     def _onchange_private_state_id(self):
         if self.private_state_id:
             self.private_country_id = self.private_state_id.country_id
 
-    @api.model
-    def get_views(self, views, options=None):
-        # Requests the My Preferences form view as last.
-        # Otherwise the fields of the 'search' view will take precedence
-        # and will omit the fields that are requested as SUPERUSER
-        # in `get_view()`.
-        preferences_view = self.env.ref("hr.res_users_view_form_preferences")
-        preferences_form = preferences_view and [preferences_view.id, 'form']
-        if preferences_form and preferences_form in views:
-            views.remove(preferences_form)
-            views.append(preferences_form)
-        result = super().get_views(views, options)
-        return result
-
-    @api.model
-    def get_view(self, view_id=None, view_type='form', **options):
-        # When the front-end loads the views it gets the list of available fields
-        # for the user (according to its access rights). Later, when the front-end wants to
-        # populate the view with data, it only asks to read those available fields.
-        # However, in this case, we want the user to be able to read/write its own data,
-        # even if they are protected by groups.
-        # We make the front-end aware of those fields by sending all field definitions.
-        # Note: limit the `sudo` to the only action of "editing own preferences" action in order to
-        # avoid breaking `groups` mecanism on res.users form view.
-        preferences_view = self.env.ref("hr.res_users_view_form_preferences")
-        if preferences_view and view_id == preferences_view.id:
-            self = self.with_user(SUPERUSER_ID)
-        result = super().get_view(view_id, view_type, **options)
-        return result
-
     @api.model_create_multi
     def create(self, vals_list):
         res = super().create(vals_list)
@@ -202,7 +179,8 @@ class ResUsers(models.Model):
         hr_fields = {
             field_name: field
             for field_name, field in self._fields.items()
-            if field.related_field and field.related_field.model_name == 'hr.employee' and field_name in vals
+            if field_name in vals
+            if callable(field.compute) and hasattr(field.compute, 'is_field_employee')
         }
 
         employee_domain = [
@@ -263,11 +241,14 @@ class ResUsers(models.Model):
         return super().action_get()
 
     @api.depends('employee_ids')
-    @api.depends_context('company')
+    @api.depends_context('company', 'uid')
     def _compute_company_employee(self):
+        Employee = self.env['hr.employee']
+        if self == self.env.user:
+            Employee = Employee.sudo()
         employee_per_user = {
             employee.user_id: employee
-            for employee in self.env['hr.employee'].search([('user_id', 'in', self.ids), ('company_id', '=', self.env.company.id)])
+            for employee in Employee.search([('user_id', 'in', self.ids), ('company_id', '=', self.env.company.id)])
         }
         for user in self:
             user.employee_id = employee_per_user.get(user)
